src/vm/eventpipeeventsource.cpp

```diff
@@ -0,0 +1,118 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#include "common.h"
+#include "eventpipeeventsource.h"
+#include "eventpipe.h"
+#include "eventpipeevent.h"
+#include "eventpipemetadatagenerator.h"
+#include "eventpipeprovider.h"
+#include "eventpipesession.h"
+
+#ifdef FEATURE_PERFTRACING
+
+const WCHAR* EventPipeEventSource::s_pProviderName = W("Microsoft-DotNETCore-EventPipe");
+const WCHAR* EventPipeEventSource::s_pProcessInfoEventName = W("ProcessInfo");
+
+EventPipeEventSource::EventPipeEventSource()
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_NOTRIGGER;
+        MODE_ANY;
+    }
+    CONTRACTL_END;
+
+    m_pProvider = EventPipe::CreateProvider(SL(s_pProviderName), NULL, NULL);
+
+    // Generate metadata.
+    const unsigned int numParams = 1;
+    EventPipeParameterDesc params[numParams];
+    params[0].Type = EventPipeParameterType::String;
+    params[0].Name = W("CommandLine");
+
+    size_t metadataLength = 0;
+    BYTE *pMetadata = EventPipeMetadataGenerator::GenerateEventMetadata(
+        1,      /* eventID */
+        s_pProcessInfoEventName,
+        0,      /* keywords */
+        0,      /* version */
+        EventPipeEventLevel::LogAlways,
+        params,
+        numParams,
+        metadataLength);
+
+    // Add the event.
+    m_pProcessInfoEvent = m_pProvider->AddEvent(
+        1,      /* eventID */
+        0,      /* keywords */
+        0,      /* eventVersion */
+        EventPipeEventLevel::LogAlways,
+        pMetadata,
+        (unsigned int)metadataLength);
+
+    // Delete the metadata after the event is created.
+    // The metadata blob will be copied into EventPipe-owned memory.
+    delete(pMetadata);
+}
+
+EventPipeEventSource::~EventPipeEventSource()
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_TRIGGERS;
+        MODE_ANY;
+    }
+    CONTRACTL_END;
+
+    // Delete the provider and associated events.
+    // This is called in the shutdown path which can't throw.
+    // Catch exceptions and ignore failures.
+    EX_TRY
+    {
+        EventPipe::DeleteProvider(m_pProvider);
+    }
+    EX_CATCH { }
+    EX_END_CATCH(SwallowAllExceptions);
+}
+
+void EventPipeEventSource::Enable(EventPipeSession *pSession)
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_ANY;
+        PRECONDITION(pSession != NULL);
+    }
+    CONTRACTL_END;
+
+    EventPipeSessionProvider *pSessionProvider = new EventPipeSessionProvider(
+        s_pProviderName,
+        -1,
+        EventPipeEventLevel::LogAlways);
+    pSession->AddSessionProvider(pSessionProvider);
+}
+
+void EventPipeEventSource::SendProcessInfo(LPCWSTR pCommandLine)
```

Comment:
I suspect it will, as we decide what other process information we want to log.  For example, we might include things like process affinity mask.  The general design is that you can easily add new parameters to the end of the event which makes things backwards compatible.  I want to avoid changing the name of this to something more specific because it means that we really need to log a new event for any new information, which just feels heavy-weight.